Blog post for producer tuning (#197)
strimzi/strimzi.github.io: _posts/2020-10-02-producer-tuning.md

### Data durability

It you want to reduce the likelihood that messages are lost, use message delivery acknowledgments.

Specify `acks=all` in your producer configuration to force a partition leader to replicate messages to a certain number of followers before acknowledging that the message request was successfully received.

```properties
# ...
acks=all
# ...
```
Use the `acks=all` producer configuration in conjunction with the `min.insync.replicas` property for topics. You set the `min.insync.replicas` property in the `KafkaTopic` resource.

The `min.insync.replicas` configuration sets the numbers of brokers that need to have logged a message before an acknowledgment is sent to the producer.

```yaml
apiVersion: kafka.strimzi.io/v1alpha1
kind: KafkaTopic
metadata:
  name: my-topic
  labels:
    strimzi.io/cluster: my-cluster
spec:
  partitions: 1
  replicas: 3
  config:
    min.insync.replicas: 2
    #...
```
Using a topic replication factor of 3, and 2 in-sync replicas on other brokers, the producer can continue unaffected if a single broker is unavailable.

> If a second broker becomes unavailable, using `acks=all` the producer won’t receive acknowledgments and won’t be able to produce more messages.

Author: I think we'll need to stick with US/global english for acknowledgement

Because of the additional checks, `acks=all` increases the latency between the producer sending a message and receiving acknowledgment. So you will have to consider the trade-off when investigating whether this is the right approach for you.

### Ordered delivery

You have two approaches to guaranteeing the order of message delivery from producers. And they depend, to a large degree, on whether or not you are using `acks=all` for data durability.

If you are using `acks=all`, you can (and should) enable idempotence for the producer to ensure that messages are delivered only once.
Using idempotence, IDs and sequence numbers are assigned to messages, so that the order is preserved even after a failed delivery.

As idempotence preserves the message ordering, you can speed the process along by increasing the number of in-flight requests you allow at one time using the `max.in.flight.requests.per.connection`.

Here we see an example configuration showing idempotence enabled, and used with `max.in.flight.requests.per.connection` and `acks=all`.

```properties
# ...
enable.idempotence=true
max.in.flight.requests.per.connection=5
acks=all
retries=2147483647
```

> The `retries` property sets the number of retries when resending a failed message request. That impressive-looking number is the default and maximum value.

Reviewer: ```suggestion
The `retries` property sets the number of retries when resending a failed message request. While that number might look impressive, what we're saying is, in effect, "retry forever".
```

Author: Thanks for clarifying. Useful.

> **Delivery timeout**
>
>If you are using `delivery.timeout.ms` in your producer configuration, producer requests will fail before the number of retries has been used if the timeout expires before a successful acknowledgment. The `delivery.timeout.ms` sets a limit on the time to wait for an acknowledgment of the success or failure to deliver a message. You can choose to leave `retries` unset and use `delivery.timeout.ms` to perform a similar function instead.

There is a performance cost to introducing additional checks to the order of delivery.

If you prefer not to use `acks=all` and idempotency,

> a lighter option is to set the number of in-flight requests to 1 (the default is 5) to preserve ordering.

Reviewer: I wouldn't use "lighter", as the lightness really depends on the requirements. Not pipelining comes with its own costs.

Author: changed to _...another option is..._

You specify a unique transactional ID in the producer configuration,
and also set the maximum allowed time for transactions in milliseconds before a timeout error is returned. The default is `900000` or 15 minutes.

```properties
# ...
enable.idempotence=true
max.in.flight.requests.per.connection=5
acks=all
retries=2147483647
transactional.id=_UNIQUE-ID_
transaction.timeout.ms=900000
# ...
```

The transactional ID is registered with the Kafka cluster on the first operation, as well as a producer `epoch` checkpoint number, which is used to identify the active producer instance.

Why might we want to identify the active producer? Say an application determines that a producer has failed and creates a new producer instance to restart a transaction. If both producers are now sending messages, duplicate records are being created and we have lost our exactly once integrity.

> By specifying a transaction ID, if a new producer instance starts, older instances of the producer are identified by their older `epoch` number and _fenced-off_ by Kafka so that their messages are no included. This maintains the integrity of the message passing by ensuring that there is only ever one valid producer with the transactional ID. Each `transactional.id` should be used for a unique set of topic partitions.

Reviewer: ```suggestion
By specifying a transaction ID, if a new producer instance starts, older instances of the producer are identified by their older `epoch` number and _fenced-off_ by Kafka so that their messages are not included. This maintains the integrity of the message passing by ensuring that there is only ever one valid producer with the transactional ID. Each `transactional.id` should be used for a unique set of topic partitions.
```

You can map topic partition names to transactional IDs, or compute the transactional ID from the topic partition names using a function that avoids collisions.

### Optimizing throughput and latency

Depending on your objective, Kafka offers a number of configuration parameters and techniques for tuning producer performance for throughput and latency.

Usually, the requirement of a system is to satisfy a particular throughput target for a proportion of messages within a given latency.

> For example, targeting 500,000 messages per second with 95% of messages being acknowledged within 2 seconds.

Reviewer: ```suggestion
For example, targeting 50,000 messages per second with 95% of messages being acknowledged within 2 seconds.
```

> * The producer is being flushed or closed

#### Compressing message batches

Compressing data batches using the `compression.type` property improves throughput and reduces the load on storage, but might not be suitable for low-latency applications where the cost of compression or decompression is prohibitive.

Message compression adds latency in the producer (CPU time spent compressing the messages),
but makes requests (and potentially disk writes) smaller, which can increase throughput.

Use the `compression.type` property to specify a valid compression codec.
You can choose `gzip`, `snappy`, `lz4`, or `zstd`, each of which have varying compression speeds.

```properties
compression.type=gzip
batch.size=32000
```

If you think compression is worthwhile, the best type of compression to use will depend on the messages being sent.

> **Adding threads**
>
> Compression is handled on the thread calling `KafkaProducer.send()`, so if the latency of this method matters for your application you can add more more threads.

#### Pipelining messages

Pipelining might sound like it has something to do with surfing the famous _Pipeline_ reef in Hawaii,

> but it's actually the equally impressive method of sending more requests from producers before the response to a previous request has been received.

Reviewer: ```suggestion
but it's actually the sending more requests from producers before the response to a previous request has been received.
```

For pipelining we use our old friend the `max.in.flight.requests.per.connection` property.
You might recall its contribution to ordered delivery earlier in this post.

Moving message requests along more frequently will obviously improve throughput. But there is a point at which you might see less beneficial effects, such as less efficient batching.

#### Adjusting message delivery waiting time

Improve throughput of your message requests by adjusting the maximum time to wait before a message is delivered and completes a send request.

Use the `delivery.timeout.ms` property to specify the maximum time in milliseconds to wait for a complete send request.
You can set the value to `MAX_LONG` to delegate to Kafka an indefinite number of retries.

You can also direct messages to a specified partition by writing a custom partitioner to replace Kafka's default, and specify the class name using the `partitioner.class` property. A custom partitioner allows you to choose how you map messages to partitions, based on the data in the message.

```properties
# ...
delivery.timeout.ms=120000
partitioner.class=my-custom-partitioner
# ...
```

### Adapt to survive

> Fine tuning your producers helps alleviate performance issues.

Reviewer: ```suggestion
Fine-tuning your producers helps alleviate performance issues.
```

But don't be tempted to make a few adjustments and think your work is done.

> You should consider fine tuning as part of a continual optimization process. Monitor Kafka regularly. Look for changing trends in usage and investigate how fine tuning can help your Kafka deployment adapt. But, as you'll know, this is only one half of the story.

Reviewer: ```suggestion
You should consider fine-tuning as part of a continual optimization process. Monitor Kafka regularly. Look for changing trends in usage and investigate how fine-tuning can help your Kafka deployment adapt. But, as you'll know, this is only one half of the story.
```

Next time we'll look at how you can optimize your consumers.

> Be sure to check back for the next installment.

Reviewer: Should we promise this?

Author: I think so. I have a _to do_ to expand the section in the guide to include consumers. Might need some help though.

author: paul_mellor
---

You can fine tune Kafka producers using configuration properties to optimize the streaming of data to consumers.

Get the tuning right, and even a small adjustment to your producer configuration can make a significant improvement to the way your producers operate.

In this post we'll discuss typical tuning considerations for Kafka producers.

<!--more-->

### Optimizing Kafka producers

Obviously, we want our producers to deliver data to Kafka topics as efficiently as possible.

But what do we mean by this, and how do we quantify it?

Do we base this on the number of messages sent over a set period of time? Or on how producers are set up to handle failure?

Before starting your adventure in optimization, think about your destination. What are the results your are hoping to achieve?

Think long enough about this, and you might find competing requirements. For example, by maximizing throughput you might also increase latency.

Be prepared to make adjustments to your adjustments.

### How is your producer performing?

It's only when you have been monitoring the performance of your producers for some time that you can gauge how best to tune their performance.

> To begin with, you might start with a basic producer configuration in development as a benchmark. When you start to analyze producer metrics to see how the producers actually perform in typical production scenarios, you can make incremental changes and make comparisons until you hit the sweet spot.

Reviewer 2: maybe mentioning the producer metrics we can put a reference to the producer monitoring documentation on the Kafka website? https://kafka.apache.org/documentation/#producer_sender_monitoring

Author: good idea. Added:

> If you want to read more about performance metrics for monitoring Kafka producers, see Kafka's [Producer Sender Metrics](https://kafka.apache.org/documentation/#producer_sender_monitoring).

When you start investigating how you to tune the performance of your producers, look at how your producers perform on average.

For example, broker restarts will have an outsize impact on very high (99%) percentile latencies. So you might concentrate on tuning your producer to achieve a latency target within a narrower bound under more typical conditions.

### Basic producer configuration

Before looking at the properties to use for fine tuning your producer,
let's assume we have a basic configuration.

Something like this.

```properties
bootstrap.servers=localhost:9092
key.serializer=org.apache.kafka.common.serialization.StringSerializer
value.serializer=org.apache.kafka.common.serialization.StringSerializer
client.id=my-client
compression.type=gzip
```

> This configuration specifies the bootstrap address for connection to the Kafka cluster, and the serializers that transform the key and value of a message.

Reviewer 2: I would be clear saying that the serializers transform the key and value from String to the corresponding raw bytes representation to be sent on the wire. Saying just "transform" doesn't sound clear to what IMHO.

Author: :+1: 

This configuration specifies the bootstrap address for connection to the Kafka cluster, and the serializers that transform the key and value of a message from from a String to its corresponding raw byte data representation.

### Data durability

It you want to reduce the likelihood that messages are lost, use message delivery acknowledgments.

Specify `acks=all` in your producer configuration to force a partition leader to replicate messages to a certain number of followers before acknowledging that the message request was successfully received.

```properties
# ...
acks=all
# ...
```
Use the `acks=all` producer configuration in conjunction with the `min.insync.replicas` property for topics. You set the `min.insync.replicas` property in the `KafkaTopic` resource.

The `min.insync.replicas` configuration sets the numbers of brokers that need to have logged a message before an acknowledgment is sent to the producer.

```yaml
apiVersion: kafka.strimzi.io/v1alpha1
kind: KafkaTopic
metadata:
  name: my-topic
  labels:
    strimzi.io/cluster: my-cluster
spec:
  partitions: 1
  replicas: 3
  config:
    min.insync.replicas: 2
    #...
```
Using a topic replication factor of 3, and 2 in-sync replicas on other brokers, the producer can continue unaffected if a single broker is unavailable.

> If a second broker becomes unavailable, the producer won’t receive acknowledgments and won’t be able to produce more messages.

Reviewer 2: It is true if the producer still sends with acks=all but it can still send if using the others type of acknowledgement.

Author: :+1: 

> If a second broker becomes unavailable, using `acks=all` the producer won’t receive acknowledgments and won’t be able to produce more messages.
